haskell-api.md

```diff
@@ -0,0 +1,236 @@
+# Haskell API Conventions
+
+## Stateless
+
+Prefer stateless endpoints. Sessions are used to authorize data access, but not
+choose what data to return.
+
+- Good:
+  - `/3/students?schools.id=x,y,z`
+  - `ApiClient.fetch("/3/students", { "schools.id": me.schoolIds })`
+- Bad:
+  - `/3/school-admins/me/students`
+  - `ApiClient.fetch("/3/school-admins/me/students")`
+
+## Parameters
+
+We want to :heart: query parameters as a way to keep the number of distinct
+endpoints lower by supporting more complex filtering combinations within each
+endpoints.
+
+- Good: `/3/students?schools.id=x,y,z&teachers.id=a,b,c`
+- Bad:
+  - `/3/schools/:id/students`
+  - `/3/teachers/:id/students`
+  - `/3/schools/:id/teachers/:id/students`
+  - etc, forever
+
+### Naming
+
+Filters should match the dotted-path of the filtered attribute in the response.
+
+For example:
+
+```
+/3/teachers
+[ { id:
+  , ...
+  , school:
+    { id:
+    , ...
+    }
+  }
+]
```

Comment:
Can you provide a sample for the request here?

```
For example for this request `/3/teachers?school.id=x` the response will be:

/3/teachers
[  { id:
    , ... 
    , school:
      { id:
      ,...
      }
  }
]
```

Reply:
I'll try to mention the request URL more prominently here. I only hesitate making this section look like it's trying to specify what the _response_ should look like, though, which I think your sentence starts to do. The section about how the response informs the query params, not the other way around.